minetest/minetest
# Port formatChatMessage to C++ for internationalization

- Goal of the PR
  Allow international strings (with 2-byte characters) in formatted chat messages.
- How does the PR work?
  1. Move `format_chat_message` functionality to the existing C++ `formatChatMessage` function.
  2. Make the `formatChatMessage` use `wstring` throughout, except for getting substrings that are only available as `string` (the rest of the `wstring` remains intact even in that case).
  3. Make an overload for getTimestamp (to allow, in this case, the format `"%H:%M:%S"` which was the format used by the Lua function).
- Does it resolve any reported issue?
  - reported by @MoNTE48  Cyrillic text is lost in chat due to calling `wide_to_narrow`, on Android 4, some other newer Android devices, and possibly a legacy OS.
- If not a bug fix, why is this PR needed? What use cases does it solve?
  - Fix bugs that started in:
    - https://github.com/minetest/minetest/pull/9133
    - https://github.com/minetest/minetest/pull/8529


## To do

This PR is Ready for Review.

- Suggested future options: If future functionality requires Lua to completely change formatChatMessage (as opposed to using only the minetest.conf setting `chat_message_format`), which seems unnecessary (and may be why format_chat_message was in Lua before, according to the vague code comment), Lua should call this function, not the other way around like before (to preserve international strings on all platforms).

## How to test
- Your computer (or Android device) must be set to Cyrillic text (Russian).
- Paste "русский текст" into chat
- Check the log
  - behavior before PR: Cyrillic characters are discarded in the GUI and in the log on some systems (verified on Android 4).
  - behavior after PR: Cyrillic characters remain intact (verified on Android 4)
